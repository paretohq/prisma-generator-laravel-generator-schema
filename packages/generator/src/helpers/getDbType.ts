import { DMMF } from "@prisma/generator-helper";

export function getDbType(field: DMMF.Field): string {
  return field.name
}