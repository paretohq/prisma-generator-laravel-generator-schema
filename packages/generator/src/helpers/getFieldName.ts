import { DMMF } from "@prisma/generator-helper";

export function getFieldName(field: DMMF.Field): string {
  return field.name
}