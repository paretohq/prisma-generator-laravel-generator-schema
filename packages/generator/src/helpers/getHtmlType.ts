import { DMMF } from "@prisma/generator-helper";

type HtmlType = 'text' | 'email' | 'number' | 'date' | 'file' | 'password' | 'select' | 'radio' | 'checkbox' | 'textarea';

export function getHtmlType(field: DMMF.Field): HtmlType {
  return 'text';
}