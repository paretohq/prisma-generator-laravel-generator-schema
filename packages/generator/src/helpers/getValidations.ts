import { DMMF } from "@prisma/generator-helper";

export function getValidations(field: DMMF.Field): String {
  const rules = [];

  if (field.isRequired)
    rules.push("required")

  // TODO: unique rules needs access to db name of model

  if (field.type === "String")
    rules.push("string")

  if (field.type === "Float" || field.type === "Decimal")
    rules.push("decimal")

  if (field.type === "Boolean")
    rules.push("boolean")

  if (field.name === "id")
    rules.push("integer|min:1")

  if (field.name === "email")
    rules.push("email")

  if (field.name === "password")
    rules.push("min:8")

  // TODO: option in custom comment to set min,max,length rules

  // TODO: option to add url rule

  // TODO: option in relationship fields to add exists rule

  // TODO: option to override inferred rules entirely

  return rules.join("|");
}