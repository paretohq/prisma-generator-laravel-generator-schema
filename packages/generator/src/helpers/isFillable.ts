import { DMMF } from "@prisma/generator-helper";

export function isFillable(field: DMMF.Field): boolean {
  return true;
}