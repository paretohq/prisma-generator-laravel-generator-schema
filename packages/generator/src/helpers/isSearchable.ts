import { DMMF } from "@prisma/generator-helper";

export function isSearchable(field: DMMF.Field): boolean {
  if (
    field.name === "name" ||
    field.name === "title" ||
    field.name === "email" ||
    field.name === "slug"
  ) return true;

  return false;
}