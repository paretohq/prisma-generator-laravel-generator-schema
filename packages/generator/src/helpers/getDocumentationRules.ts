import { Rules } from "./getValidations";

function splitRules(rules: string): Array<string> {
  const rules_string = rules.split("(")[1].split(")")[0];
  return rules_string.split("|").map(rule_string => `@rule(${rule_string})`)
}

export function getDocumentationRules(documentation: string | undefined): [boolean, Rules] {
  const ruleDefinitions: Array<string> = [];
  const attributes = documentation?.split(" ") ?? [];
  let override = false;

  for (let attribute of attributes) {
    if (attribute.startsWith("@rules")) {
      ruleDefinitions.push(...splitRules(attribute));
      override = true;
      break;
    }

    if (attribute.startsWith("@rule"))
      ruleDefinitions.push(attribute)
  }

  return [
    override,
    ruleDefinitions.reduce((result: Rules, rd) => {
      rd = rd.split("(")[1].split(")")[0];
      const [k, v] = rd.split(":");
      result[k] = [v]
      return result;
    }, {})
  ]
}