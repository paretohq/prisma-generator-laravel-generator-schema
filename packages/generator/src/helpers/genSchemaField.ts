import { DMMF } from '@prisma/generator-helper';
import { getDbType } from "./getDbType";
import { getFieldName } from "./getFieldName";
import { getHtmlType } from "./getHtmlType";
import { getValidations } from "./getValidations";
import { isFillable } from "./isFillable";
import { isInForm } from "./isInForm";
import { isInIndex } from "./isInIndex";
import { isInView } from "./isInView";
import { isPrimary } from "./isPrimary";
import { isSearchable } from "./isSearchable";

export const genSchemaField = (field: DMMF.Field) => ({
  name: getFieldName(field),
  dbType: getDbType(field),
  htmlType: getHtmlType(field),
  validations: getValidations(field),
  searchable: isSearchable(field),
  fillable: isFillable(field),
  primary: isPrimary(field),
  inForm: isInForm(field),
  inIndex: isInIndex(field),
  inView: isInView(field)
})