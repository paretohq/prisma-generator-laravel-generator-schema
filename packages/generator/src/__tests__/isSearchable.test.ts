import { isSearchable } from "../helpers/isSearchable"
import { getSampleDMMF } from './__fixtures__/getSampleDMMF'

test('is searchable', async () => {
  const sampleDMMF = await getSampleDMMF()

  sampleDMMF.datamodel.models.forEach((model) => {
    model.fields.forEach(field => {
      expect(isSearchable(field)).toMatchSnapshot()
    })
  })
})
