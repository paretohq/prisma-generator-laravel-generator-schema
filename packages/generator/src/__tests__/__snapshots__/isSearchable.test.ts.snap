// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is searchable 1`] = `false`;

exports[`is searchable 2`] = `true`;

exports[`is searchable 3`] = `true`;
